chainerx_cc/chainerx/backward.cc

```diff
@@ -621,13 +639,28 @@ std::vector<absl::optional<Array>> Grad(
         }
     }
 
-    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map)}.Run();
+    // Push initial output grads, Run assigns to 1 in other case
+    for (size_t i = 0; i < grad_outputs.size(); i++) {
+        const std::shared_ptr<ArrayBody>& array_body = internal::GetArrayBody(outputs[i]);
+        if (const std::shared_ptr<ArrayNode>& array_node = array_body->GetArrayNode(actual_backprop_id)) {
+            output_grads.emplace_back(grad_outputs[i]);
+            array_node_grad_map.emplace(array_node.get(), internal::GradRef{&output_grads.back()});
+        }
+    }
+
+    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map), retain_grad}.Run();
 
+    size_t i = 0;
```

Comment:
Why using different index than that of `input_grads`?
(In other words in which case `grad.has_value()` evaluates to false?)

Reply:
grad.has_value() evaluates to false in the case that the corresponding input value was an array that is not part of the graph.

I rewrote the loop to be in the same style of the above ones and iterate input_grads and inputs with the same index.

Reply:
Comment updated with more specific content